torchrl/collectors/collectors.py

```diff
@@ -124,6 +128,28 @@ def _get_policy_and_device(
                     "env must be provided to _get_policy_and_device if policy is None"
                 )
             policy = RandomPolicy(self.env.action_spec)
+        elif (
+            observation_spec is not None
+            and isinstance(policy, nn.Module)
+            and not isinstance(policy, TensorDictModule)
```

Comment:
I guess we could write a helper function to determine if a policy is not reading tensordicts
Here are the possible cases:
- the policy is a TensorDictModule
- the policy is a nn.Module that has in_keys, out_keys and reads tensordicts but it is not a TensorDictModule
- the policy is a nn.Module that reads tensors (and maybe also tensordicts)
In the last case we should wrap it into a TensorDictModule

So what we could do is:
- if it's a TDModule we're good
- If it's not but it has in_keys, out_keys and the signature reads only one value we're good
- Otherwise we should wrap that thing in a TensorDictModule

Another option would be to look at the type-hinting in the forward of the module but I'm not sure we really need that

Reply:
Yes, I think this makes sense.

When you say helper function, do you mean pulling the `elif` condition out into a call to e.g. `_module_is_tensordict_compatible(module)` or similar? Or did you mean the actual body of the `elif`?

Reply:
Had a go at this, let me know if it's what you had in mind.